### PR TITLE
Add GLM4 model

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

Adds GLM model.

### CODE REVIEW

File: src/transformers/models/glm/configuration_glm.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+#           🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
+#               This file was automatically generated from <path_to_modular_file.py>.
+#         Do NOT edit this file manually as any edits will be overwritten by the generation of
+#         the file from the modular. If any change should be done, please apply the change to the
+#                           modular_xxx.py file directly. One of our CI enforces this
+#           🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
+# coding=utf-8
+# Copyright 2024 Google Inc. HuggingFace Inc. team. All rights reserved.
```
Comment (ArthurZucker) [inline src/transformers/models/glm/configuration_glm.py:8]: probably wrong company!

File: src/transformers/models/glm/configuration_glm.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from ...configuration_utils import PretrainedConfig
+
+
+class GlmConfig(PretrainedConfig):
+    model_type = "glm"
```
Comment (ArthurZucker) [inline src/transformers/models/glm/configuration_glm.py:28]: docstring is missing!

File: src/transformers/models/glm/configuration_glm.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        post_layer_norm=True,
+        use_cache=True,
+        tie_word_embeddings=False,
+        rope_theta=10000.0,
+        pad_token_id=151329,
+        eos_token_id=[151329, 151336, 151338],
+        bos_token_id=None,
+        head_dim=128,
+        attention_bias=True,
+        linear_bias=False,
```
Comment (ArthurZucker) [inline src/transformers/models/glm/configuration_glm.py:55]: all of this seems similar to llama / gemma

File: src/transformers/models/glm/configuration_glm.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        num_attention_heads=32,
+        num_key_value_heads=2,
+        resid_pdrop=0.0,
+        attention_dropout=0.0,
+        hidden_act="silu",
+        max_position_embeddings=131072,
+        initializer_range=0.02,
+        rms_norm_eps=0.00000015625,
+        use_rms_norm=True,
+        apply_residual_connection_post_layernorm=False,
```
Comment (ArthurZucker) [inline src/transformers/models/glm/configuration_glm.py:45]: is this false for all models? If so, to delete!

File: src/transformers/models/glm/configuration_glm.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/glm/configuration_glm.py:55]: -> meaing you can inherit from llama and update the defaults that are different / add the new ones

File: src/transformers/models/glm/modular_glm.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+        super().__init__()
+
+        self.config = config
+        self.apply_residual_connection_post_layernorm = config.apply_residual_connection_post_layernorm
+        self.self_attn = GLM_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx=layer_idx)
+
+        self.mlp = GlmMLP(config)
+        self.input_layernorm = (
+            GlmRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+            if config.use_rms_norm
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:442]: check what config uses, but we avoid that in general as well! (code path)

File: src/transformers/models/glm/modular_glm.py [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+            position_embeddings (`Tuple[torch.FloatTensor, torch.FloatTensor]`, *optional*):
+                Tuple containing the cosine and sine positional embeddings of shape `(batch_size, seq_len, head_dim)`,
+                with `head_dim` being the embedding dimension of each attention head.
+            kwargs (`dict`, *optional*):
+                Arbitrary kwargs to be ignored, used for FSDP and other methods that injects code
+                into the model
+        """
+
+        hidden_states_after_norm = self.input_layernorm(hidden_states)
+        residual = hidden_states_after_norm if self.apply_residual_connection_post_layernorm else hidden_states
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:490]: same here! check if any released models have both

File: src/transformers/models/glm/modular_glm.py [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+        return outputs
+
+
+class GlmModel(LlamaModel):
+    def __init__(self, config: GlmConfig):
+        super().__init__(config)
+        self.layers = nn.ModuleList(
+            [GlmDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        if config.post_layer_norm:
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:529]: same here

File: src/transformers/models/glm/__init__.py
Comment (ArthurZucker) [inline src/transformers/models/glm/__init__.py:1]: Init should be like this: https://github.com/huggingface/transformers/pull/31329/files#diff-e13de4b5db0c6872b5f0ec197d07fdaf80174b37f55bd4fefbe1526e57635683 
(you could not know but we ought to enforce this now!)

**Review by ArthurZucker — COMMENTED** [review]
Super nice, you are missing the test files, integration tests etc! (And readme etc)

File: src/transformers/models/glm/modular_glm.py [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., 0::2]
+    x2 = x[..., 1::2]
+    return torch.stack((-x2, x1), dim=-1).flatten(-2)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:118]: I think the issue is that here the unsqeeze dim is different!

Comment (ArthurZucker) [reply src/transformers/models/glm/modular_glm.py:118]:
```

def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):

```
The converter probably indexes the full function with the arguments!

File: src/transformers/models/glm/modular_glm.py [hunk 9]
```diff
@@ -0,0 +1,575 @@
...
+
+
+GLM_ATTENTION_CLASSES = {
+    "eager": GlmAttention,
+    "flash_attention_2": GlmFlashAttention2,
+    "sdpa": GlmSdpaAttention,
+}
+
+
+class GlmDecoderLayer(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:458]: this one looks fairly classic, I would have supposed you don't need the forward (unless the issue is with the name of layers?)

**Review by ArthurZucker — COMMENTED** [review]
Something went wrong with the rebasing / merging as you have unrelated changes!

**Comment by Cyrilvallez** [comment]
> Something went wrong with the rebasing / merging as you have unrelated changes!

Yes, currently looking at it

**Comment by ArthurZucker** [comment]
Superseed by #33823
